### PR TITLE
Add support for named argument functions

### PR DESCRIPTION
**Pull request by eyalleshem**
Add support for named argument function

This commit support functions with argument names.
the format is :
"Select some_function( a => exp, b => exp2 .. ) FROM table1
OR
"select * from table(function(a => exp)) f;"

see:
https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#named-argument-assignment-token
or example is snwflake db :
https://docs.snowflake.com/en/sql-reference/functions/flatten.html
 This commit does not belong to any branch on this repository.

### CODE REVIEW

**Comment by eyalleshem** [comment]
Sorry about some style changes that made by `cargo fmt`

**Comment by nickolay** [comment]
I find it weird that you added NamedArgument as a new variant of Expr. This makes the parser accept many weird constructs, which shouldn't be accepted. I expected this to be a new SelectItem-like enum. You can use `peek_nth_token(1) == RArrow` to determine if it's a named argument.

The argument name is a simple identifier, not an ObjectName.

Please check the PR for typos (FUCNTION, rigth, agument, // vs /// (which led to rustfmt reformatting the whole enum))

> the format is : "Select * from ( a => exp, b=> exp2 .. )

Surely you meant `SELECT myfunc(args)` rather than `FROM (args)`

**Comment by eyalleshem** [comment]
Sorry for all the typos , I fixed them (Hope that all of them..) 
I also changed the the object name to ident - and fixed the commit message with the correct query . 

But i don't sure what you means when you say that the named-argument should be `SelectItem` 
It's a comma-separated argument list that passed to function, In the `struct Function` the args is comma-separated `Vec<Expr>` , And because that my thought was that creating new `Expr` type will be the best suit .

**Comment by nickolay** [comment]
I said a _new_ enum, a little like SelectItem (currently with two variants: unnamed argument, named argument).

**Comment by eyalleshem** [comment]
I done the changes according to comments .

File: src/ast/mod.rs [hunk 1]
```diff
@@ -873,12 +873,42 @@ impl fmt::Display for Assignment {
...
+    pub name: Ident,
+    pub arg: Expr,
+}
+
+impl fmt::Display for FunctionNamedArg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} => {}", self.name, self.arg)?;
+        Ok(())
+    }
+}
```
Comment (nickolay) [inline src/ast/mod.rs:888]: Is there a reason to make this a separate struct, rather than `FunctionArg::Named { name, arg }`?

_(Another option is to make the argument type a simple struct with an optional `name`, and the reason I suggested `enum` initially because of the [other possible kinds of SQL arguments](https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#SQL-argument)...)_

File: src/ast/mod.rs [hunk 2]
```diff
@@ -873,12 +873,42 @@ impl fmt::Display for Assignment {
...
+pub enum FunctionArg {
+    Named(FunctionNamedArg),
+    Unnamed(Expr),
+}
+
+impl fmt::Display for FunctionArg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            FunctionArg::Named(named_arg) => named_arg.fmt(f),
+            FunctionArg::Unnamed(unnamed_arg) => unnamed_arg.fmt(f),
```
Comment (nickolay) [inline src/ast/mod.rs:901]: We use `write!(f, "{}", ...)` in other places for this.

File: src/parser.rs [hunk 3]
```diff
@@ -2201,11 +2201,24 @@ impl Parser {
         Ok(Assignment { id, value })
     }
 
-    pub fn parse_optional_args(&mut self) -> Result<Vec<Expr>, ParserError> {
+    fn parse_function_args(&mut self) -> Result<FunctionArg, ParserError> {
+        if self.peek_nth_token(1) == Token::RArrow {
+            let name = self.parse_identifier()?;
+            // the `=>`
+            let _ = self.next_token();
```
Comment (nickolay) [inline src/parser.rs:2208]: Please use `expect_token` instead.

File: src/tokenizer.rs [hunk 4]
```diff
@@ -323,7 +326,7 @@ impl<'a> Tokenizer<'a> {
                         }
                     }
                 }
-                // identifier or keyword
+                // identifier or Keyword
```
Comment (nickolay) [inline src/tokenizer.rs:329]: Why this change?

File: src/ast/mod.rs [hunk 5]
```diff
@@ -141,7 +141,7 @@ impl fmt::Display for ObjectName {
 
 /// An SQL expression of any type.
 ///
-/// The parser does not distinguish between expressions of different types
+/// The parser does not distinguish between qessions of different types
```
Comment (nickolay) [inline src/ast/mod.rs:144]: Typo?

File: src/ast/mod.rs (hunk 1 above)
Comment (eyalleshem) [reply src/ast/mod.rs]: Fixed , 
I agree that enum is better.

File: src/ast/mod.rs (hunk 2 above)
Comment (eyalleshem) [reply src/ast/mod.rs]: fixed

File: src/parser.rs (hunk 3 above)
Comment (eyalleshem) [reply src/parser.rs]: fixed

File: src/tokenizer.rs (hunk 4 above)
Comment (eyalleshem) [reply src/tokenizer.rs]: sorry ,  by mistake ..

File: src/ast/mod.rs (hunk 5 above)
Comment (eyalleshem) [reply src/ast/mod.rs]: sorry ,  by mistake ..

**Comment by nickolay** [comment]
Thanks!
